internal/cfgmgr/cfgmgr.go

```
@@ -0,0 +1,149 @@
+	DEVPKEY_LOCATIONPATHS_GUID = "a45c254e-df1c-4efd-8020-67d146a850e0"
+
+	STRING_BUFFER_SIZE = uint64(1024)
+)
+
+func getDevPKeyDeviceLocationPaths() (*devPropKey, error) {
+	guid, err := guid.FromString(DEVPKEY_LOCATIONPATHS_GUID)
+	if err != nil {
+		return nil, err
+	}
+	return &devPropKey{
+		fmtid: guid,
+		pid:   37,
+	}, nil
+}
+
+func GetDeviceLocationPathsFromIDs(ids []string) ([]string, error) {
+	result := []string{}
+	devPKeyDeviceLocationPaths, err := getDevPKeyDeviceLocationPaths()
+	if err != nil {
+		return nil, err
+	}
+	for _, id := range ids {
+		var devNodeInst uint32
+		err = cmLocateDevNode(&devNodeInst, id, CM_LOCATE_DEVNODE_NORMAL)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to locate device node for %s", id)
+		}
+		propertyType := uint64(0)
+		propertyBufferSize := STRING_BUFFER_SIZE
+		var propertyBuffer [STRING_BUFFER_SIZE]uint16
+		err = cmGetDevNodeProperty(devNodeInst, devPKeyDeviceLocationPaths, &propertyType, &propertyBuffer[0], &propertyBufferSize, 0)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get location path property from device node for %s with", id)
+		}
+		if propertyType != DEVPROP_TYPE_STRING_LIST {
+			return nil, fmt.Errorf("expected to return property type DEVPROP_TYPE_STRING_LIST %d, instead got %d", DEVPROP_TYPE_STRING_LIST, propertyType)
+		}
+		if propertyBufferSize > STRING_BUFFER_SIZE {
+			return nil, fmt.Errorf("location path is too long, size %d", propertyBufferSize)
+		}
+		formattedResult := convertFirstNullTerminatedValueToString(propertyBuffer[:propertyBufferSize])
+		result = append(result, formattedResult)
+	}
+
+	return result, nil
+}
+
+// helper function that finds the first \u0000 rune and returns the wide string as a regular go string
+func convertFirstNullTerminatedValueToString(buf []uint16) string {
+	r := utf16.Decode(buf)
+	converted := string(r)
+	zerosIndex := strings.IndexRune(converted, '\u0000')
+	if zerosIndex != -1 {
```

Is there any valid case where the null terminator will be missing from the buffer?

---

Not that I can think of, just this was just cautionary since we don't control the data that's returned which we attempt to convert here.

---

In that case we should return an error when `zerosIndex == -1`, rather than returning the entire buffer as a string, as this case indicates malformed data has been passed to us.

---

Done.

---

This check needs to be `==` instead of `!=`.